### PR TITLE
Fabric

### PR DESCRIPTION
**Pull request by RogerioTeixeira**
Add definition class WebglFilterBackend and Canvas2dFilterBackend

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/fabric/fabric-impl.d.ts [hunk 1]
```diff
@@ -4480,3 +4481,50 @@ interface IUtil extends IUtilAnimation, IUtilArc, IObservable<IUtil>, IUtilDomEv
 	object: IUtilObject;
 	string: IUtilString;
 }
+
+export interface FilterBackend {
+	resources: Object;
+
+	applyFilters(filters: IBaseFilter[], sourceElement: HTMLImageElement | HTMLCanvasElement, sourceWidth: number, sourceHeight: number, targetCanvas: HTMLCanvasElement): any;
+
+	evictCachesForKey(cacheKey: any): void;
```
Comment (NotWoods) [inline types/fabric/fabric-impl.d.ts]: Should `cacheKey` be [string](https://github.com/kangax/fabric.js/blob/1e5c64ff2ae1f7c32367f8a64616d9bd6919d425/src/shapes/image.class.js#L163)?

File: types/fabric/fabric-impl.d.ts [hunk 2]
```diff
@@ -4480,3 +4481,50 @@ interface IUtil extends IUtilAnimation, IUtilArc, IObservable<IUtil>, IUtilDomEv
...
+
+	chooseFastestCopyGLTo2DMethod(width: number, height: number): void;
+
+	createWebGLCanvas(width: number, height: number): void;
+
+	applyFiltersDebug(filters: IBaseFilter[], sourceElement: HTMLImageElement | HTMLCanvasElement, sourceWidth: number, sourceHeight: number, targetCanvas: HTMLCanvasElement, cacheKey: any): any;
+
+	glErrorToString(context: any, errorCode: any): string;
+
+	createTexture(gl: WebGLRenderingContext, width: number, height: number, textureImageSource: HTMLImageElement | HTMLCanvasElement): WebGLTexture;
```
Comment (NotWoods) [inline types/fabric/fabric-impl.d.ts]: `textureImageSource` seems to be [optional](https://github.com/kangax/fabric.js/blob/1fa3fdae31d2ee9a1c7c4bebb975d66b3171b5e9/src/filters/base_filter.class.js#L172)

File: types/fabric/fabric-impl.d.ts [hunk 3]
```diff
@@ -4480,3 +4481,50 @@ interface IUtil extends IUtilAnimation, IUtilArc, IObservable<IUtil>, IUtilDomEv
 	object: IUtilObject;
 	string: IUtilString;
 }
+
+export interface FilterBackend {
+	resources: Object;
```
Comment (NotWoods) [inline types/fabric/fabric-impl.d.ts]: `Object` refers to a fabric Object, but from what I can tell [`resources`](https://github.com/kangax/fabric.js/blob/1fa3fdae31d2ee9a1c7c4bebb975d66b3171b5e9/src/filters/2d_backend.class.js#L26) is a plain Javascript object map. If so, use an indexed type like `{ [key: string]: HTMLCanvasElement }`.

File: types/fabric/fabric-impl.d.ts (hunk 2 above)
Comment (NotWoods) [reply types/fabric/fabric-impl.d.ts]: There's also another overload seen [here](https://github.com/kangax/fabric.js/blob/75ba9e1ac82a36fa28aed3660c90896ba864f017/src/filters/blendimage_filter.class.js#L107)

File: types/fabric/fabric-impl.d.ts [hunk 4]
```diff
@@ -4480,3 +4481,50 @@ interface IUtil extends IUtilAnimation, IUtilArc, IObservable<IUtil>, IUtilDomEv
...
+export let filterBackend: FilterBackend;
+export interface Canvas2dFilterBackend extends FilterBackend { }
+export class Canvas2dFilterBackend {
+	constructor();
+}
+
+export interface WebglFilterBackendOptions {
+	tileSize: number;
+}
+export interface WebglFilterBackend extends FilterBackend, WebglFilterBackendOptions {
```
Comment (NotWoods) [inline types/fabric/fabric-impl.d.ts]: WebglFilterBackend's `applyFilters` takes an optional `cacheKey` parameter (equivalent parameters to `applyFiltersDebug`

File: types/fabric/fabric-impl.d.ts [hunk 5]
```diff
@@ -4480,3 +4481,50 @@ interface IUtil extends IUtilAnimation, IUtilArc, IObservable<IUtil>, IUtilDomEv
...
+
+	createWebGLCanvas(width: number, height: number): void;
+
+	applyFiltersDebug(filters: IBaseFilter[], sourceElement: HTMLImageElement | HTMLCanvasElement, sourceWidth: number, sourceHeight: number, targetCanvas: HTMLCanvasElement, cacheKey: any): any;
+
+	glErrorToString(context: any, errorCode: any): string;
+
+	createTexture(gl: WebGLRenderingContext, width: number, height: number, textureImageSource: HTMLImageElement | HTMLCanvasElement): WebGLTexture;
+
+	getCachedTexture(uniqueId: any, textureImageSource: HTMLImageElement | HTMLCanvasElement): WebGLTexture;
```
Comment (NotWoods) [inline types/fabric/fabric-impl.d.ts]: `uniqueId` should be [`string`](https://github.com/kangax/fabric.js/blob/master/src/filters/webgl_backend.class.js#L312)

**Review by NotWoods — CHANGES_REQUESTED** [review]
Tests look good! I poked around in the source code and there are some areas that don't seem to match how the functions are used internally.

**Comment by RogerioTeixeira** [comment]
Thanks for your review !! 
Can you please look at the new code and update your review status if appropriate?

**Review by NotWoods — APPROVED** [review]
Looks good to me 👍

**Pull request merged**